src/core/tactics.ml

```diff
@@ -8,65 +8,128 @@ open Pos
+    let goal_has_no_meta_value = function
+      | Goal.Unif _ -> true
+      | Goal.Typ gt -> (
+          match !(gt.goal_meta.meta_value) with
+          | Some _ -> false
+          | None -> true
+        )
+    in
+    let gs_typ = List.filter goal_has_no_meta_value gs_typ in
+    {ps with proof_goals = new_gs_unif @ gs_typ}
+  with
+  | Unif.Unsolvable ->
+    fatal pos "Constraints are unsolvable !"
+
 (** [handle_tactic ss ps tac] tries to apply the tactic [tac] (in the proof
      state [ps]), and returns the new proof state.  This function fails
      gracefully in case of error. *)
-let handle_tactic : Sig_state.t -> Proof.t -> p_tactic -> Proof.t =
-  fun ss ps tac ->
+let handle_tactic :
+  Sig_state.t -> Terms.expo -> Proof.t -> p_tactic -> Proof.t =
+  fun ss e ps tac ->
   (* First handle the tactics that do not change the goals. *)
   match tac.elt with
   | P_tac_print         ->
       (* Just print the current proof state. *)
       Console.out 1 "%a" pp_goals ps; ps
   | P_tac_proofterm     ->
-      (* Just print the current proof term. *)
-      let t = Meta(ps.proof_term, [||]) in
-      let name = ps.proof_name.elt in
-      Console.out 1 "Proof term for %s: %a\n" name pp_term t; ps
+    begin
+      match ps.proof_term with
+      | Some proof_term ->
+        (* Just print the current proof term. *)
+        let t = Meta(proof_term, [||]) in
+        let name = ps.proof_name.elt in
+        Console.out 1 "Proof term for %s: %a\n" name pp_term t; ps
+      | None ->
+        Console.out 1 "No proof term"; ps
+    end
   | P_tac_query(q)      ->
       Queries.handle_query ss (Some ps) q; ps
   | _                   ->
   (* The other tactics may change the goals. *)
   (* Get the focused goal and the other goals. *)
-  let (g, gs) =
-    match ps.proof_goals with
-    | []    -> fatal tac.pos "There is nothing left to prove.";
-    | g::gs -> (g, gs)
+    if ps.proof_goals = [] then
+      fatal tac.pos "There is nothing left to prove.";
+
+  match tac.elt with
+  | P_tac_print
+  | P_tac_proofterm
+  | P_tac_query(_)      -> assert false (* Handled above. *)
+  | P_unif_solve        -> solve ps tac.pos
+  | _                   ->
+
+  (* Get the unif goals, the first type goal and the following goals *)
+  let pre_g, g, post_g =
+    let rec first_goal_typ pre post =
+      match post with
+      | [] -> pre,None,post
+      | (Goal.Typ gt) :: post -> pre,Some(gt),post
+      | (Goal.Unif _ as gu) :: post -> first_goal_typ (pre @ [gu]) post
+    in
+    first_goal_typ [] ps.proof_goals
   in
-  (* Obtaining the goal environment and type. *)
-  let (env, a) = Goal.get_type g in
 
-  let scope = Scope.scope_term Public ss env in
-  let infer t = Typing.infer (Env.to_ctxt env) t in
-  let check t a = Typing.check (Env.to_ctxt env) t a in
+  (* Obtaining the goal environment and type. *)
+  let scope t =
+    let (env, _a) = match g with
+      | Some gt -> Goal.get_type gt
+      | None -> assert false
+    in
```

Comment:
It seems to me that the computation of env, which does not depend on t, should be factorized out of the scope function. This would simplify many calls to scope where env is not used.

Reply:
Fixed in https://github.com/Deducteam/lambdapi/pull/422/commits/a93d56890b92d4a60a7b2b0f6a9281c9a635fb91